File: sec5/p054.py
n = int(input())
w, k = map(int, input().split())
INF = 10 ** 9
l = [INF] + list(map(int, input().split()))
dp = [[INF]*(w+1) for _ in range(n+1)]
dp[0][0] = 0
for i in range(1, n+1):
    for j in range(w+1):
        if j < l[i]:
            dp[i][j] = dp[i-1][j]
            continue
        dp[i][j] = min(dp[i-1][j], dp[i-1][j-l[i]]+1)
if dp[i][j] <= k:
    print('Yes')
else:
    print('No')

File: sec5/p054.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

var scan = bufio.NewScanner(os.Stdin)

const (
	initBufSize = 10000
	maxBufSize = 10000000
)

func init() {
	buf := make([]byte, initBufSize)
	scan.Buffer(buf, maxBufSize)
	scan.Split(bufio.ScanWords)
}

func main() {
	n := NumStdin()
	w := NumStdin()
	k := NumStdin()

	a := make([]int, n)
	for i := 0; i < n; i++ {
		a[i] = NumStdin()
	}

	dp := make([][]int, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = make([]int, w+1)
		for j := 0; j <= w; j++ {
			dp[i][j] = math.MaxInt64
		}
	}
	dp[0][0] = 0

	for i := 0; i < n; i++ {
		for j := 0; j <= w; j++ {
			chmin(&dp[i+1][j], dp[i][j])
			
			if a[i] <= j && dp[i][j-a[i]] < math.MaxInt64 {
				chmin(&dp[i+1][j], dp[i][j-a[i]] + 1)
			}
		}
	}

	if dp[n][w] <= k {
		fmt.Println("Yes")
	} else {
		fmt.Println("No")
	}
}

func NumStdin() int {
	scan.Scan()
	num, err := strconv.Atoi(scan.Text())
	if err != nil {
		panic(err)
	}
	return num
}

func chmin(a *int, b int) {
	if *a > b {
		*a = b
	}
}
